### PR TITLE
Add Devart Oracle provider

### LINKED ISSUE #3585: add an oracle driver?
Is it possible to add a Devart.Data.Oracle driver？ thx

### LINKED ISSUE #3601: Oracle schema datatype issue - REF CURSOR is not handled
Oracle schema datatype issue - REF CURSOR is not handled

### Patch
```diff
diff --git a/Source/LinqToDB/DataProvider/Oracle/OracleSchemaProvider.cs b/Source/LinqToDB/DataProvider/Oracle/OracleSchemaProvider.cs
index 46ef2a5c8..2aa2ecd58 100644
--- a/Source/LinqToDB/DataProvider/Oracle/OracleSchemaProvider.cs
+++ b/Source/LinqToDB/DataProvider/Oracle/OracleSchemaProvider.cs
@@ -548,6 +550,7 @@ protected override DataType GetDataType(string? dataType, string? columnType, in
 				case "VARCHAR2"               : return DataType.VarChar;
 				case "XMLTYPE"                : return DataType.Xml;
 				case "ROWID"                  : return DataType.VarChar;
+				case "REF CURSOR"             : return DataType.Cursor;
 				default:
 					if (dataType?.StartsWith("TIMESTAMP") == true)
 						return dataType.EndsWith("TIME ZONE") ? DataType.DateTimeOffset : DataType.DateTime2;
```

### Environment details
linq2db version: latest
Database Server: Oracle v18
Database Provider: Oracle Managed
Operating system: Win10
.NET Framework: 4.7.2

### PR DESCRIPTION
**Pull request by MaceWindu**
Fix #3585 
Fix #3601

- add Devart provider
- add Oracle 18 and 21 to CI

New provider is not tested on CI due to licensing limitations

### CODE REVIEW

**Comment by MaceWindu** [comment]
And now my opinion on provider.

If you don't plan to use features specific to this provider - use `Oracle.ManagedDataAccess`:
- no async support in 2022
- several things working with Oracle providers, doesn't work (discovered by our tests):
  - disappearing `\0` character
  - some numbers could lose precision, e.g. `100` could be returned as `99.999...` and then rounded to `99`
  - issues with xml, binary_double and binary_float types
  - multiple issues with procedure parameters binding

**Comment by jods4** [comment]
> no async support in 2022

For the record official managed ODP.NET has no async support either, it's implemented as sync-over-async.
Oracle announced a true async implementation for 2021, now they say it's gonna be in Q4 2022. 🤞

**Comment by MaceWindu** [comment]
@jods4 it is understandable. Oracle is a small company with limited resources. I think they will add async to Oracle provider after they add it to `MySql.Data` provider :pray:

**Comment by MaceWindu** [comment]
OK, at least on CI everything works as expected (could be client-server issue with olson vs windows timezone db). Anyway, will try to fix it, but in general it is ready.

File: Source/LinqToDB/Expressions/TypeMapper/TypeMapper.cs [hunk 1]
```diff
@@ -49,19 +49,21 @@ public void RegisterTypeWrapper(Type wrapperType, Type originalType)
...
 
-			if (wrapperType.Name != originalType.Name)
+			var wrapperAttr = wrapperType.GetCustomAttribute<WrapperAttribute>(true);
+
+			if ((wrapperAttr?.TypeName ?? wrapperType.Name) != originalType.Name)
 				throw new LinqToDBException($"Original and wraped types should have same type name. {wrapperType.Name} != {originalType.Name}");
-			if (_types.ContainsKey(originalType.Name))
-				throw new LinqToDBException($"Type with name {originalType.Name} already registered in mapper");
+			if (_types.ContainsKey(originalType.FullName ?? originalType.Name))
+				throw new LinqToDBException($"Type with name {originalType.FullName} already registered in mapper");
```
Comment (viceroypenguin) [inline Source/LinqToDB/Expressions/TypeMapper/TypeMapper.cs:57]: If `originalType.FullName` can be null, then the exception may have incomplete information.

File: Tests/Linq/Linq/DateTimeFunctionsTests.cs [hunk 2]
```diff
@@ -1268,9 +1268,10 @@ public void AddYearFromColumnExpression([DataSources(TestProvName.AllInformix)]
 					from t in db.Types select Sql.AsSql(Sql.DateDiff(Sql.DateParts.Hour, t.DateTimeValue, t.DateTimeValue.AddHours(100))));
 		}
 
+		// Devart: returns 100 as 99.999...
 		[Test]
```
Comment (viceroypenguin) [inline Tests/Linq/Linq/DateTimeFunctionsTests.cs:1272]: Would it make sense to use `[ActiveIssue("Devart returns 100 as 99.999", TestProvName.AllOracleDevart)]` here?

**Review by viceroypenguin — DISMISSED** [review]

**Review by viceroypenguin — APPROVED** [review]

**Review by sdanyliv — APPROVED** [review]

**Pull request merged**